arch/arm/core/aarch32/thread.c

```diff
@@ -274,7 +274,12 @@ uint32_t z_check_thread_stack_fail(const uint32_t fault_addr, const uint32_t psp
 #if defined(CONFIG_USERSPACE)
 	if (thread->arch.priv_stack_start) {
 		/* User thread */
-		if ((__get_CONTROL() & CONTROL_nPRIV_Msk) == 0U) {
+#if defined(CONFIG_CPU_CORTEX_R)
+		if (arch_is_user_context())
```

Comment:
Would `if (arch_is_user_context())` work for all cases, not just Cortex-R?

Reply:
arch_is_user_context() doesn't work for cortex-M.  I tried it and it broke the mem_protect/userspace test on the mps2_an385 board.  arch_is_user_context() on Cortex-M has an extra check of IPSR that means this function does not just mean privileged vs user mode.  Instead, I created a new function call arch_is_unprivileged_state function that just checks the privilege level.  This should maintain the same level of functionality for both M and R, but abstracts the difference into a header file.